cmd/scan_txs.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+   Copyright 2020 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include <absl/container/flat_hash_set.h>
+#include <absl/flags/flag.h>
+#include <absl/flags/parse.h>
+#include <absl/flags/usage.h>
+#include <absl/time/time.h>
+
+#include <boost/filesystem.hpp>
+#include <iostream>
+#include <silkworm/db/access_layer.hpp>
+#include <silkworm/execution/execution.hpp>
+
+ABSL_FLAG(std::string, datadir, silkworm::db::default_path(), "chain DB path");
+ABSL_FLAG(uint64_t, from, 1, "start from block number (inclusive)");
+ABSL_FLAG(uint64_t, to, UINT64_MAX, "check up to block number (exclusive)");
+
+int main(int argc, char* argv[]) {
+    absl::SetProgramUsageMessage("Executes Ethereum blocks and compares resulting change sets against DB.");
+    absl::ParseCommandLine(argc, argv);
+
+    namespace fs = boost::filesystem;
+
+    fs::path db_path(absl::GetFlag(FLAGS_datadir));
+    if (!fs::exists(db_path) || !fs::is_directory(db_path) || db_path.empty()) {
+        std::cerr << absl::GetFlag(FLAGS_datadir) << " does not exist.\n";
+        std::cerr << "Use --db flag to point to a Turbo-Geth populated chaindata.\n";
+        return -1;
+    }
+
+    using namespace silkworm;
+
+    lmdb::options db_opts{};
+    db_opts.read_only = true;
+    std::shared_ptr<lmdb::Environment> env{lmdb::get_env(absl::GetFlag(FLAGS_datadir).c_str(), db_opts)};
+
+    // counters
+    uint64_t nTxs{0}, nErrors{0};
+
+    const uint64_t from{absl::GetFlag(FLAGS_from)};
+    const uint64_t to{absl::GetFlag(FLAGS_to)};
+
+    uint64_t block_num{from};
+    for (; block_num < to; ++block_num) {
+        std::unique_ptr<lmdb::Transaction> txn{env->begin_ro_transaction()};
```

Comment:
No need to open a new transaction for each block.
Just use one immediately after the db opening. this will save one abort and one reopen (with new id assignment) for each loop

Reply:
Thanks so much for your comments.

Two quick questions: 

I read on the LMDB documentation not to create a long-lived transaction. The process I'm running runs for hours, so I thought that would be long-lived. But now that I think about it, I think I read about 'reset' as well. Should I use a 'reset' in each loop?

This code was copied from `./check_changes.cpp`. Should I change it there too?

Reply:
Yes long lived ro transactions are discouraged but only when they're interleaved with rw transactions: an opened ro transaction may prevent rw transactions from reusing free pages in free_dbi.
If, such in this case, the procedure is meant to quickly go through blocks (using a dedicated db - i.e. TG is not running) a single transaction saves a lot of processing time.

Reply:
WRT the `reset` keyword. Is not actually a method of transaction itself rather the release of the shared_ptr.
Basically reset on a shared ptr invokes the destructor of the pointed class (RAII).

Reply:
TG is running on the same machine in my case. Just to be safe, I think I'll leave it the way it is. I can write a comment.

Question: if there is a long-lasting read transaction while TG is running and the database 'grows quickly' as the docs say, does the growth correct itself over time, or is the file much larger forever (ignoring migrations or compactions)?

One of the things I wanted to do with this code is make it very simple. I'd like some day to write a tutorial on scanning the chain for data.

Reply:
The `growth corrects itself` under these circumstances which must both be true:
* new data is inserted in "upsert" mode (i.e. not using MDB_APPEND or MDB_APPENDDUP flags)
* the ongoing transaction can reclaim free pages from a list of pages recorded in free_dbi where the transaction id (of the freed pages) is **lower** than the id of the current transaction.

Reply:
> TG is running on the same machine in my case. Just to be safe,

Be aware it takes a lot longer

Reply:
I actually ran the tool with both a single long-lasting database transaction and a per-loop transaction, and I found no obvious difference. In fact, in some cases, the per-loop database transaction was faster. Here's some numbers:

https://docs.google.com/spreadsheets/d/13omV5sfpKa6dI-2nB7PgbHEqtKvymCpdcZifv1TlDlU/edit#gid=461497860

See the timings tab.